src/h.js

```diff
@@ -1,22 +1,29 @@
-export function h(tag, data) {
-  var node
-  var stack = []
-  var children = []
+function addChild(children, val) {
+  if (val == null || typeof val === 'boolean') return
 
-  for (var i = arguments.length; i-- > 2; ) {
-    stack[stack.length] = arguments[i]
+  if (Array.isArray(val)) {
+    for (var j = 0; j < val.length; j++) {
```

Comment:
Well I'm really sorry to insist but that's probably wrong and oversimplified (please read this excellent article by M. Egorov, V8 and Java VM guy).

```js
 var i=0, len=data.length;
  for (; i < len; i++) {
    (6 + 1);
  }
```

This loop does nothing. It's well explained in the video I linked above. `(6+1)` is removed by [loop invariant code motion](https://www.wikiwand.com/en/Loop-invariant_code_motion). The loop being empty, it's removed too thanks to [dead code elimination](https://www.wikiwand.com/en/Dead_code_elimination), ...and `i` and `len` assignments too.
So the above piece of code simply disappears after compilation and you end up measuring the outer benchmark loop, and only that 😉 

From M. Egorov:
> Don’t blindly trust microbenchmarks, they are often not measuring what you think they are measuring.

Like I mentioned in my previous comment, benchmarks are not deterministic in time and results may vary because of external factors (OS process scheduling, GC pauses, and many more). `457 / 440` is basically a 1% variation which is a way too small threshold to make any serious conclusion.
I even bet you sometimes get swapped digits like in the example I linked above, where non-cached was faster. And that's for one simple reason: they are doing the same thing.

JIT compilers are really complex, and depending on certain circumstances they optimize things or not. In the cases you actually got positive results, and if really related to `length` property, chances are that the length couldn't be cached by the compiler because it couldn't tell if the array would be mutated or not, so it keeps the access property subroutine.
In the bench you provided me (and in `h`), it's obvious to the compiler that the array will never be mutated, so it safely caches it for us (cf. the article I posted above). The results speaks for themselves: there is not any measurable difference in the benchmarks (and I do 1M loops, not 1K).

But then there is more, some excerpts from M. Egorov's article:

> Thing to realize here is simple: if compiler is unable to figure out that arr.length does not change within the loop, you can’t help it much by manually caching it. When it needs to access array element, compiler anyway has to emit a bounds check which requires loading length.

> Ok. We can’t reason about caching arr.length as an optimization that improves the code by reducing a number of arr.length accesses. [...] Well, in fact we do make some thing somewhat worse: the life of a poor register allocator. When we manually cache length we give it one more live value to care about and keep around in the loop.

> Understand before acting [...] Yes, surprisingly it has nothing to do with caching or not caching array.length access. Don’t derive your code style from microbenchmarks. Don’t base it on advices from esteemed elders (or mine for that matter). Instead invest into understanding where, how and why your application is spending CPU cycles.

Conclusion: do not optimize something until you analyzed compilation artifacts with a tool like IRHydra2 and highlighted sections of [SSA](https://www.wikiwand.com/fr/Static_single_assignment_form) that proves the compiler didn't do a good job.

Every piece of `h` was benched and analyzed with IRHydra2 to track deopts and optimal SSA. I'm not saying that the current version is perfect, it's not. But I did invest some time measuring stuff precisely, and `length` did not show up

Until I have some solid benchmark and traces that actually prove there is a bug in the current `v8` implementation under `h` circumstances (and I would be highly interested in it btw), I'll stick to this conclusion.